epersongroups.md

```diff
@@ -0,0 +1,310 @@
+
+Filter using parameter:
+* all : retrieve indirect groups as well
+
+Samples:
+* GET /rest/api/eperson/groups/3b1de75d-5bf9-4ca3-bf4d-42bc8abd0d42/groups
+* GET /rest/api/eperson/groups/3b1de75d-5bf9-4ca3-bf4d-42bc8abd0d42/groups?all=true
+
+
+```json
+{
+    "_embedded": {
+      "groups": [
+        {
+          "id": "617cf46b-535c-42d5-9d22-327ce2eff6dc",
+          "uuid": "617cf46b-535c-42d5-9d22-327ce2eff6dc",
+          "name": "Administrator",
+          "handle": null,
+          "metadata": [
+            {
+              "key": "dc.title",
+              "value": "Administrator",
+              "language": null
+            }
+          ],
+          "permanent": true,
+          "type": "group",
+          "_links": {
+            "groups": {
+              "href": "https://dspace7.4science.it/dspace-spring-rest/api/eperson/groups/617cf46b-535c-42d5-9d22-327ce2eff6dc/groups"
+            },
+            "epersons": {
+              "href": "https://dspace7.4science.it/dspace-spring-rest/api/eperson/groups/617cf46b-535c-42d5-9d22-327ce2eff6dc/epersons"
+            },
+            "self": {
+              "href": "https://dspace7.4science.it/dspace-spring-rest/api/eperson/groups/617cf46b-535c-42d5-9d22-327ce2eff6dc"
+            }
+          },
+          "_embedded": {
+            "groups": {
+              "_embedded": {
+                "groups": []
+              },
+              "_links": {
+                "self": {
+                  "href": "https://dspace7.4science.it/dspace-spring-rest/api/eperson/groups/617cf46b-535c-42d5-9d22-327ce2eff6dc/groups"
+                }
+              },
+              "page": {
+                "number": 0,
+                "size": 0,
+                "totalPages": 1,
+                "totalElements": 0
+              }
+            }
+          }
+        }
+      ]
+    },
+    "_links": {
+      "self": {
+        "href": "https://dspace7.4science.it/dspace-spring-rest/api/eperson/groups/3b1de75d-5bf9-4ca3-bf4d-42bc8abd0d42/groups"
+      }
+    },
+    "page": {
+      "number": 0,
+      "size": 10,
+      "totalPages": 1,
+      "totalElements": 10
+    }
+}
+```
+
+### Add a Group to a parent Group
+
+**POST /api/eperson/groups/<:parentgroupuuid>/groups**
+
+To add a child group to a parent group, perform a POST to the Sub groups of a Group endpoint when logged in as admin.
+
+The actual sub group is part of the body using the uri-list
+
+Example:
+```bash
+curl -i -X POST "https://dspace7.4science.it/dspace-spring-rest/api/eperson/groups/617cf46b-535c-42d5-9d22-327ce2eff6dc/groups"
+  -H "Content-Type:text/uri-list"
+  -d "https://dspace7.4science.it/dspace-spring-rest/api/eperson/groups/05e3dbb8-332b-4487-a3f9-d78431b6cc02"
+```
+
+The group is mandatory
+
+Return codes:
+* 204: if the update succeeded 
+* 401 Forbidden - if you are not authenticated
+* 403 Unauthorized - if you are not logged in with sufficient permissions 
+* 422: if the specified group is not found
```

Comment:
it could be useful to state 404 explicitly to discriminate between the case that the "parent" group is not found and the 422 child group is not found

What if the child cannot be added to the parent due to a cyclic reference (ie. A -> B -> C, -> mean contains we cannot add A to C)? 409 Conflict could be appropriate here. It will be ideal if we report to the client that A should be removed of B to be added to C

Reply:
404 and 409 can indeed be useful
@abollini can you elaborate on how it should be further reported to the client that A should be removed of B to be added to C? Do you have an example for this

Reply:
We can start with just reporting that as plain text, it is similar to what we do when an eperson cannot be deleted due to db constraints.
Something like: The group A cannot be added as it is a parent of group C via group B will be ideal but we can also just start with "The group cannot be added due to a cyclic reference"

BTW, looking for an example I found that EPerson delete should be treated in the same way, see for instance
https://github.com/DSpace/DSpace/blob/master/dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/EPersonRestRepository.java#L249

If we like to go with 409 I suggest to open a JIRA ticket to fix the above contract/implementation 

the Item deletions are different 
https://github.com/DSpace/DSpace/blob/master/dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/ItemRestRepository.java#L196
https://github.com/DSpace/DSpace/blob/master/dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/ItemRestRepository.java#L200

as the client cannot "solve" the conflicts (if it is a templateItem or an inprogress submission it needs to use a different endpoint, there will be no a second call to the delete after "solving").

Reply:
I've included it in this contract.
For the others a JIRA ticket will indeed be required